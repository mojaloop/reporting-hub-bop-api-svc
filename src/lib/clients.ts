/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

import { Collection, MongoClient } from 'mongodb';
import Config from './config';
import { ConnectionString } from 'connection-string';
import { logger } from '../shared/logger';

const createMongoClient = (): MongoClient => {
  const csMongoDBObj = new ConnectionString();
  csMongoDBObj.setDefaults({
    protocol: 'mongodb',
    hosts: [{ name: Config.EVENT_STORE_DB.HOST, port: Config.EVENT_STORE_DB.PORT }],
    user: Config.EVENT_STORE_DB.USER,
    password: Config.EVENT_STORE_DB.PASSWORD,
    path: [Config.EVENT_STORE_DB.DATABASE],
    params: Config.EVENT_STORE_DB.PARAMS,
  });
  const mongoUri = csMongoDBObj.toString();
  const safeUri = mongoUri.replace(/(\/\/)(.*):(.*)@/, '$1****:****@');
  logger.info(`Connecting to MongoDB with URI: ${safeUri}`);
  const mongoClient = new MongoClient(mongoUri);
  mongoClient.connect();
  return mongoClient;
};

const createEventStoreClient = (): Collection => {
  const mongoClient = createMongoClient();
  const db = mongoClient.db(Config.EVENT_STORE_DB.DATABASE);
  return db.collection('reporting');
};

const createTransactionClient = (): Collection => {
  const mongoClient = createMongoClient();
  const db = mongoClient.db(Config.EVENT_STORE_DB.DATABASE);
  return db.collection('transaction');
};

const createSettlementClient = (): Collection => {
  const mongoClient = createMongoClient();
  const db = mongoClient.db(Config.EVENT_STORE_DB.DATABASE);
  return db.collection('settlement');
};

export { createTransactionClient, createEventStoreClient, createSettlementClient };
