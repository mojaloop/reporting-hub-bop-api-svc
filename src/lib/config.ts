import Convict from 'convict';
import path from 'path';

// interface to represent service configuration
export interface ServiceConfig {
  env: string;
  PORT: number;
  REPORTING_DB: {
    DIALECT: string;
    HOST: string;
    PORT: number;
    USER: string;
    PASSWORD: string;
    SCHEMA: string;
  };
  EVENT_STORE_DB: {
    HOST: string;
    PORT: number;
    USER: string;
    PASSWORD: string;
    DATABASE: string;
  };
  ORY_KETO_READ_URL: string;
  AUTH_CHECK_PARTICIPANTS: boolean;
  USER_ID_HEADER: string;
  PRISMA_LOGGING_ENABLED: boolean;
  CORS_WHITELIST: string[];
  ALLOW_CREDENTIALS: boolean;
}

// Declare configuration schema, default values and bindings to environment variables
export const ConvictConfig = Convict<ServiceConfig>({
  env: {
    doc: 'The application environment.',
    format: ['default', 'production', 'development', 'test', 'integration', 'e2e'],
    default: 'default',
    env: 'NODE_ENV',
  },
  PORT: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3000,
    env: 'PORT',
  },
  REPORTING_DB: {
    DIALECT: {
      doc: 'DIALECT',
      format: '*',
      default: 'mysql',
      env: 'REPORTING_DB_DIALECT',
    },
    HOST: {
      doc: 'The Hostname/IP address of database',
      format: '*',
      default: 'localhost',
      env: 'REPORTING_DB_HOST',
    },
    PORT: {
      doc: 'The port of database',
      format: 'port',
      default: 3306,
      env: 'REPORTING_DB_PORT',
    },
    USER: {
      doc: 'The username for database',
      format: '*',
      default: 'central_ledger',
      env: 'REPORTING_DB_USER',
    },
    PASSWORD: {
      doc: 'The password for database',
      format: '*',
      default: 'password',
      env: 'REPORTING_DB_PASSWORD',
    },
    SCHEMA: {
      doc: 'The schema in database',
      format: '*',
      default: 'central_ledger',
      env: 'REPORTING_DB_SCHEMA',
    },
  },
  EVENT_STORE_DB: {
    HOST: {
      doc: 'The Hostname/IP address of database',
      format: '*',
      default: 'localhost',
      env: 'EVENT_STORE_DB_HOST',
    },
    PORT: {
      doc: 'The port number of database',
      format: 'port',
      default: 27017,
      env: 'EVENT_STORE_DB_PORT',
    },
    USER: {
      doc: 'The user of database',
      format: '*',
      default: 'test',
      env: 'EVENT_STORE_DB_USER',
    },
    PASSWORD: {
      doc: 'The password of database',
      format: '*',
      default: 'test123',
      env: 'EVENT_STORE_DB_PASSWORD',
    },
    DATABASE: {
      doc: 'The database name in database',
      format: '*',
      default: 'admin',
      env: 'EVENT_STORE_DB_DATABASE',
    },
  },
  ORY_KETO_READ_URL: {
    doc: 'The read URL of Ory Keto',
    format: '*',
    default: '',
    env: 'ORY_KETO_READ_URL',
  },
  AUTH_CHECK_PARTICIPANTS: {
    doc: 'AUTH_CHECK_PARTICIPANTS',
    format: 'Boolean',
    default: false,
    env: 'AUTH_CHECK_PARTICIPANTS',
  },
  USER_ID_HEADER: {
    doc: 'USER_ID_HEADER',
    format: '*',
    default: 'x-user',
    env: 'USER_ID_HEADER',
  },
  PRISMA_LOGGING_ENABLED: {
    doc: 'PRISMA_LOGGING_ENABLED',
    format: 'Boolean',
    default: false,
    env: 'PRISMA_LOGGING_ENABLED',
  },
  CORS_WHITELIST: {
    doc: 'CORS_WHITELIST',
    format: Array,
    default: [],
    env: 'CORS_WHITELIST',
  },
  ALLOW_CREDENTIALS: {
    doc: 'ALLOW_CREDENTIALS',
    format: 'Boolean',
    default: false,
    env: 'ALLOW_CREDENTIALS',
  },
});

// Load environment dependent configuration
const env = ConvictConfig.get('env');
const configFile = process.env.CONFIG_FILE || path.join(__dirname, `../../config/${env}.json`);
ConvictConfig.loadFile(configFile);

// Perform configuration validation
ConvictConfig.validate({ allowed: 'strict' });

// extract simplified config from Convict object
const config: ServiceConfig = {
  env: ConvictConfig.get('env'),
  PORT: ConvictConfig.get('PORT'),
  REPORTING_DB: ConvictConfig.get('REPORTING_DB'),
  EVENT_STORE_DB: ConvictConfig.get('EVENT_STORE_DB'),
  ORY_KETO_READ_URL: ConvictConfig.get('ORY_KETO_READ_URL'),
  AUTH_CHECK_PARTICIPANTS: ConvictConfig.get('AUTH_CHECK_PARTICIPANTS'),
  USER_ID_HEADER: ConvictConfig.get('USER_ID_HEADER'),
  PRISMA_LOGGING_ENABLED: ConvictConfig.get('PRISMA_LOGGING_ENABLED'),
  CORS_WHITELIST: ConvictConfig.get('CORS_WHITELIST'),
  ALLOW_CREDENTIALS: ConvictConfig.get('ALLOW_CREDENTIALS'),
};

export default config;
