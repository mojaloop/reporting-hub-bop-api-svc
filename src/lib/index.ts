/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

export * from './prisma';
export * from './auth';
export * from './mongo';
export * from './requestInfoParser';
