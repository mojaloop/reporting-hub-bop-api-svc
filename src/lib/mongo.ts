/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

import { Collection, MongoClient } from 'mongodb';

let collection: Collection;
let client: MongoClient | null = null;

// Not being used anywhere right now
const getMongoClient = async (uri: string, collectionName: string): Promise<Collection> => {
  if (!collection) {
    client = new MongoClient(uri);
    await client.connect();
    collection = client.db().collection(collectionName);
  }
  return collection;
};

const closeMongoClientConnection = async (): Promise<void> => {
  if (client) {
    await client.close();
    client = null;
  }
};

export { getMongoClient, Collection, closeMongoClientConnection };
