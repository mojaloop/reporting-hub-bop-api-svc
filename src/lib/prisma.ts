/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

import { PrismaClient as CentralLedgerClient } from '@app/generated/centralLedger';
import { PrismaClient as EventStoreClient } from '@app/generated/eventStore';
import Config from './config';
import { ConnectionString } from 'connection-string';

const createCentralLedgerClient = (logQuery = false): CentralLedgerClient => {
  const csMysqlObj = new ConnectionString();
  csMysqlObj.setDefaults({
    protocol: 'mysql',
    hosts: [{ name: Config.REPORTING_DB.HOST, port: Config.REPORTING_DB.PORT }],
    user: encodeURIComponent(Config.REPORTING_DB.USER),
    password: encodeURIComponent(Config.REPORTING_DB.PASSWORD),
    path: [Config.REPORTING_DB.SCHEMA],
  });

  return new CentralLedgerClient({
    datasources: {
      db: {
        url: csMysqlObj.toString(),
      },
    },
    log: logQuery ? ['query'] : [],
  });
};

const createEventStoreClient = (logQuery = false): EventStoreClient => {
  const csMongoDBObj = new ConnectionString();
  csMongoDBObj.setDefaults({
    protocol: 'mongodb',
    hosts: [{ name: Config.EVENT_STORE_DB.HOST, port: Config.EVENT_STORE_DB.PORT }],
    user: encodeURIComponent(Config.EVENT_STORE_DB.USER),
    password: encodeURIComponent(Config.EVENT_STORE_DB.PASSWORD),
    path: [Config.EVENT_STORE_DB.DATABASE],
  });
  return new EventStoreClient({
    datasources: {
      db: {
        url: csMongoDBObj.toString(),
      },
    },
    log: logQuery ? ['query'] : [],
  });
};

export { createCentralLedgerClient, createEventStoreClient, CentralLedgerClient, EventStoreClient };
