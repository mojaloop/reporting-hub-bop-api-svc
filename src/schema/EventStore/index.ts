/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

import Query from './Query';
import Event from './Event';

export default [Query, Event];
