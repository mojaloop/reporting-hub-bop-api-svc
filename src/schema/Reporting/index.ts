const mongoose = require('mongoose');
const Schema = mongoose.Schema;

// Define the schema for the payer and payee parties
const PartySchema = new Schema({
  partyIdType: String,
  partyIdentifier: String,
  partyName: String,
  supportedCurrencies: String,
});

// Define the schema for the quote request
const QuoteRequestSchema = new Schema({
  quoteId: String,
  amountType: String,
  amount: {
    currency: String,
    amount: Number,
  },
  fees: {
    currency: String,
    amount: Number,
  },
});

// Define the schema for the transfer terms
const TransferTermsSchema = new Schema({
  transferAmount: {
    currency: String,
    amount: Number,
  },
  payeeReceiveAmount: {
    currency: String,
    amount: Number,
  },
  payeeFspFee: {
    currency: String,
    amount: Number,
  },
  payeeFspCommission: {
    currency: String,
    amount: Number,
  },
  expiration: Date,
  geoCode: {
    latitude: String,
    longitude: String,
  },
  ilpPacket: String,
});

// Define the schema for the conversion terms
const ConversionTermsSchema = new Schema({
  conversionId: String,
  determiningTransferId: String,
  initiatingFsp: String,
  counterPartyFsp: String,
  amountType: String,
  sourceAmount: {
    currency: String,
    amount: Number,
  },
  targetAmount: {
    currency: String,
    amount: Number,
  },
  expiration: Date,
  charges: [
    {
      chargeType: String,
      sourceAmount: {
        currency: String,
        amount: Number,
      },
      targetAmount: {
        currency: String,
        amount: Number,
      },
    },
  ],
  ilpPacket: String,
});

// Define the schema for conversions
const ConversionSchema = new Schema({
  conversionRequestId: String,
  conversionId: String,
  conversionCommitRequestId: String,
  conversionState: String,
  conversionStateChanges: [
    {
      conversionState: String,
      dateTime: Date,
      reason: String,
    },
  ],
  counterPartyFSP: String,
  conversionType: String,
  conversionSettlementWindowId: Schema.Types.Int64,
  conversionTerms: ConversionTermsSchema,
});

// Define the schema for transactions
const TransactionSchema = new Schema({
  transferId: { type: String, index: true },
  transactionId: String,
  sourceAmount: Number,
  sourceCurrency: { type: String, index: true },
  targetAmount: Number,
  targetCurrency: { type: String, index: true },
  createdAt: { type: Date, index: true },
  lastUpdated: Date,
  transferState: { type: String, index: true },
  transferStateChanges: [
    {
      transferState: String,
      dateTime: Date,
      reason: String,
    },
  ],
  transactionType: { type: String, index: true },
  errorCode: { type: String, index: true },
  transferSettlementWindowId: Schema.Types.Int64,
  payerDFSP: { type: String, index: true },
  payerDFSPProxy: { type: String, index: true },
  payeeDFSP: { type: String, index: true },
  payeeDFSPProxy: { type: String, index: true },
  positionChanges: [
    {
      participantName: String,
      currency: String,
      ledgerType: String,
      dateTime: Date,
      updatedPosition: String,
      change: String,
    },
  ],
  payerParty: {
    partyIdType: { type: String, index: true },
    partyIdentifier: { type: String, index: true },
    partyName: String,
    supportedCurrencies: String,
  },
  payeeParty: {
    partyIdType: { type: String, index: true },
    partyIdentifier: { type: String, index: true },
    partyName: String,
    supportedCurrencies: String,
  },
  quoteRequest: QuoteRequestSchema,
  transferTerms: TransferTermsSchema,
  conversions: [ConversionSchema],
});

// Define the schema for settlement windows
const SettlementWindowSchema = new Schema({
  settlementWindowId: { type: Schema.Types.Int64, index: true },
});

// Define the schema for settlements
const SettlementSchema = new Schema({
  settlementWindows: [SettlementWindowSchema],
  settlementId: { type: Schema.Types.Int64, index: true },
  createdAt: Date,
  lastUpdatedAt: Date,
  settlementModel: String,
  settlementStatus: String,
});

// Create models
const Transaction = mongoose.model('Transaction', TransactionSchema);
const Settlement = mongoose.model('Settlement', SettlementSchema);

module.exports = {
  Transaction,
  Settlement,
};
export default [Transaction, Settlement];
