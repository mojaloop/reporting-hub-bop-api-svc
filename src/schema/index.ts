/**************************************************************************
 *  (C) Copyright Mojaloop Foundation 2020                                *
 *                                                                        *
 *  This file is made available under the terms of the license agreement  *
 *  specified in the corresponding source code repository.                *
 *                                                                        *
 *  ORIGINAL AUTHOR:                                                      *
 *       Yevhen Kyriukha <yevhen.kyriukha@modusbox.com>                   *
 **************************************************************************/

import 'tsconfig-paths/register';
import { makeSchema, mutationType, queryType } from 'nexus';
import { join, dirname } from 'path';

import Scalar from './Scalar';
import Transaction from './Transfer';
import TransactionSummary from './TransferSummary';

const types = [Scalar, Transaction, TransactionSummary];

const Query = queryType({
  definition(t) {
    t.field('_placeholder', {
      type: 'Boolean',
    });
  },
});

const Mutation = mutationType({
  definition(t) {
    t.field('_placeholder', {
      type: 'Boolean',
    });
  },
});

export default makeSchema({
  types: [Query, Mutation, types.flat()],
  outputs: process.env.NODE_ENV !== 'production' && {
    typegen: join(dirname(require.resolve('@app/index')), '../node_modules/@types/nexus-typegen/index.d.ts'),
    schema: join(dirname(require.resolve('@app/index')), '../node_modules/@types/nexus-typegen/schema.graphql'),
  },
  contextType: {
    module: '@app/context',
    export: 'Context',
  },
  sourceTypes: {
    modules: [
      {
        module: '@prisma/client',
        alias: 'prisma',
      },
    ],
  },
});
