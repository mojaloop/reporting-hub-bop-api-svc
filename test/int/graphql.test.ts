import { graphql } from 'graphql';
import { getMongoClient, Collection, closeMongoClientConnection } from '@app/lib';

const testTransfers = require('./data/reporting.transfers.json');
import schema from '../../src/schema';

describe('TransferSummary Integration Tests', () => {
  const MONGO_URI = process.env.MONGO_URI || 'mongodb://localhost:27017';
  let collection: Collection;

  beforeAll(async () => {
    try {
      collection = await getMongoClient(MONGO_URI, 'transfer');
      await collection.deleteMany({});

      console.log('Test database setup completed');
    } catch (error) {
      console.error('Database setup failed:', error);
      throw error;
    }
  });

  afterAll(async () => {
    try {
      if (collection) {
        await collection.deleteMany({});
        await closeMongoClientConnection();
      }
      console.log('Test database cleanup completed');
    } catch (error) {
      console.error('Database cleanup failed:', error);
      throw error;
    }
  });

  describe('insert data into live MongoDB', () => {
    it('should insert transfer summary data', async () => {
      // Insert test data
      const result = await collection.insertMany(testTransfers);

      const insertedIds = result.insertedIds;

      // get inserted ids count
      const insertedIdsCount = Object.keys(insertedIds).length;

      // Assert
      expect(insertedIdsCount).toEqual(1);
      // Only 1 row in sample mock data so expects 1 at the moment
    });
  });
});

describe('Send sample graphql queries', () => {
  let mockCtx;
  beforeEach(() => {
    const mockTransfer = {
      transferId: 'b51ec534-ee48-4575-b6a9-ead2955b8069',
      transactionId: 'transaction123',
      sourceAmount: 100,
      sourceCurrency: 'USD',
      targetAmount: 90,
      targetCurrency: 'EUR',
      createdAt: '2024-11-14T06:24:56.719Z',
      lastUpdated: '2024-11-14T06:24:56.720Z',
      transferState: 'COMPLETED',
      transactionType: 'TRANSFER',
    };

    mockCtx = {
      transaction: {
        transaction: {
          findUnique: jest.fn().mockResolvedValue(mockTransfer),
          findMany: jest.fn().mockResolvedValue([mockTransfer]),
        },
      },
    };
  });

  it('should send a transfer query', async () => {
    const query = `
      query {
        transfer(transferId: "b51ec534-ee48-4575-b6a9-ead2955b8069") {
          transferId
          transactionId
          sourceAmount
          sourceCurrency
          targetAmount
          targetCurrency
          createdAt
          lastUpdated
          transferState
          transactionType
        }
      }
    `;
    const result = await graphql({
      schema,
      source: query,
      contextValue: mockCtx,
    });

    // Assert the expected result
    expect(result).toMatchSnapshot();
    expect(mockCtx.transaction.transaction.findUnique).toHaveBeenCalledWith({
      where: { transferId: 'b51ec534-ee48-4575-b6a9-ead2955b8069' },
    });
  });
});
