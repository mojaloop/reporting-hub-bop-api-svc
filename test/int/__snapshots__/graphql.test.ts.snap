// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send sample graphql queries should send a transfer query 1`] = `
Object {
  "data": Object {
    "transfer": Object {
      "createdAt": "2024-11-14T06:24:56.719Z",
      "lastUpdated": "2024-11-14T06:24:56.720Z",
      "sourceAmount": 100,
      "sourceCurrency": "USD",
      "targetAmount": 90,
      "targetCurrency": "EUR",
      "transactionId": "transaction123",
      "transactionType": "TRANSFER",
      "transferId": "b51ec534-ee48-4575-b6a9-ead2955b8069",
      "transferState": "COMPLETED",
    },
  },
}
`;
